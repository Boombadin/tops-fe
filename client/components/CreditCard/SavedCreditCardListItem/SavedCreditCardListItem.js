import React from 'react';
import styled from 'styled-components';

import { TextGuide } from '@client/components/Typography';
import withLocales from '@client/hoc/withLocales';

import {
  ELEMENT_ACTION,
  ELEMENT_TYPE,
  generateTestId,
} from '../../../utils/generateElementId';

const CardBox = styled.div`
  flex: 0 0 100%;
`;
const CardBoxInside = styled.div`
  display: flex;
  min-height: 48px;
  padding: 0 15px;
  font-size: 13px;
  position: relative;
`;
const CardInfo = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
`;
const CardInfoBank = styled.div`
  display: flex;
  align-items: center;
  flex: 1;
  min-width: 0;
`;
const CardDefault = styled.div`
  width: 16px;
  height: 16px;
  flex: 0 0 auto;
  margin-right: 14px;
`;
const CardDefaultIcon = styled.img`
  width: 16px;
  height: 16px;
`;

const CardTypeIcon = styled.div`
  width: 32px;
  height: 20px;
  display: flex;
  align-items: center;
  margin-right: 17px;
  & img {
    max-width: 32px;
  }
`;
const BankName = styled(TextGuide)`
  color: #171722;
  margin-right: 5px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;
const DefaultText = styled(TextGuide)`
  flex: 0 0 auto;
`;
const CardNumber = styled.div`
  flex: 0 0 auto;
  margin-left: 15px;
`;

const SavedCreditCardListItem = ({
  translate,
  card,
  isCanBeSelect = false,
  selectedCard,
  bankName,
  onSelectCard = () => {},
}) => {
  const isCardSelected = isCanBeSelect ? selectedCard?.id === card?.id : false;
  if (bankName === '') {
    bankName = '';
  } else if (bankName === null || bankName === 'undefined') {
    bankName = translate('payment.credit_card.other_card');
  }
  return (
    <CardBox
      data-testid={generateTestId({
        type: ELEMENT_TYPE.INFO,
        action: ELEMENT_ACTION.VIEW,
        moduleName: 'SavedCreditCardListItem',
        uniqueId: 'Container',
      })}
    >
      <CardBoxInside
        onClick={() => {
          onSelectCard(card);
        }}
        data-testid={generateTestId({
          type: ELEMENT_TYPE.BUTTON,
          action: ELEMENT_ACTION.EDIT,
          moduleName: 'SavedCreditCardListItem',
          uniqueId: 'OnSetSelectCard',
        })}
      >
        <CardInfo>
          <CardInfoBank>
            <CardDefault>
              {isCardSelected && (
                <CardDefaultIcon
                  data-testid={generateTestId({
                    type: ELEMENT_TYPE.IMAGE,
                    action: ELEMENT_ACTION.VIEW,
                    moduleName: 'SavedCreditCardListItem',
                    uniqueId: 'DefaultCard',
                  })}
                  src="/assets/icons/red-checked.svg"
                />
              )}
            </CardDefault>
            <CardTypeIcon>
              <img src={card?.typeIcon} />
            </CardTypeIcon>
            <BankName
              data-testid={generateTestId({
                type: ELEMENT_TYPE.INFO,
                action: ELEMENT_ACTION.VIEW,
                moduleName: 'SavedCreditCardListItem',
                uniqueId: 'BankName',
              })}
              type="body"
            >
              {bankName}
            </BankName>
            {card?.isDefault && (
              <DefaultText type="body" color={'#b4b8c1'}>
                {translate('payment.credit_card.card_list_default')}
              </DefaultText>
            )}
          </CardInfoBank>
          <CardNumber>
            {card?.maskedNumberByStar.substr(
              card?.maskedNumberByStar.length - 6,
              6,
            )}
          </CardNumber>
        </CardInfo>
      </CardBoxInside>
    </CardBox>
  );
};

export default withLocales(SavedCreditCardListItem);
