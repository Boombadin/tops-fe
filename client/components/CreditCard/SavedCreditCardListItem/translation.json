{
  "th_TH": {
    "edit": "แก้ไข",
    "delete": "ลบ",
    "default_card": "บัตรชำระเงินเริ่มต้น",
    "expired_date": "วันหมดอายุ",
    "expired": "EXPIRED"
  },
  "en_US": {
    "edit": "Edit",
    "delete": "Delete",
    "default_card": "Default Payment Card",
    "expired_date": "Expires",
    "expired": "EXPIRED"
  }
}
