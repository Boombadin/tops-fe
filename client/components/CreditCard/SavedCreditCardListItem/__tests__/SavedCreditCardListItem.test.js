import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';

import SavedCreditCardListItem from '../SavedCreditCardListItem';

jest.mock('@client/hoc/withLocales', () => {
  return Component => props => (
    <Component translate={jest.fn(key => key)} {...props} />
  );
});

const TEST_ID = {
  SELECT_CARD: 'btn-editSavedCreditCardListItem-OnSetSelectCard',
  DEFAULT_CARD: 'img-viewSavedCreditCardListItem-DefaultCard',
  BANK_NAME: 'inf-viewSavedCreditCardListItem-BankName',
};

describe('components/CreditCard/SavedCreditCardListItem', () => {
  // mock component props exclude translate
  const props = {
    card: {
      bankId: null,
      expiryMonth: 8,
      expiryYear: 2023,
      id: '592358235717763073',
      isDefault: true,
      isExpired: false,
      masked_number: '555555XXXXXX4444',
      maskedNumberByStar: '**** **** **** 4444',
      type: 'Visa',
      typeIcon: '/icons/types-credit-card/master-no-border.svg',
      bankImage: '/images/payment/product-img-empty.svg',
      bankName: '',
    },
    isSelected: false,
    selectedCard: '',
  };

  test(`it should match with snapshot`, () => {
    const { asFragment } = render(<SavedCreditCardListItem {...props} />);
    expect(
      asFragment(<SavedCreditCardListItem {...props} />),
    ).toMatchSnapshot();
  });

  test('Bank name match with remote config | it should be show bank name', () => {
    const props = {
      ...props,
      bankName: 'Bank of Ayudhya',
    };
    render(<SavedCreditCardListItem {...props} />);

    expect(screen.getByText('Bank of Ayudhya')).toBeInTheDocument();
  });

  test('Bank name = null | it should be show other bank', () => {
    const props = {
      ...props,
      bankName: null,
    };
    render(<SavedCreditCardListItem {...props} />);
    expect(
      screen.getByText('payment.credit_card.other_card'),
    ).toBeInTheDocument();
  });

  test('Bank name = empty | it should be show empty name', () => {
    const props = {
      ...props,
      bankName: '',
    };
    const { getByTestId } = render(<SavedCreditCardListItem {...props} />);
    const getBankName = getByTestId(TEST_ID.BANK_NAME);
    expect(getBankName).toBeEmptyDOMElement;
  });

  test('Click selected card | it should be call function onSelectCard()', () => {
    const newProps = {
      ...props,
      isCanBeSelect: true,
      onSelectCard: () => {},
    };
    const { getByTestId } = render(<SavedCreditCardListItem {...newProps} />);
    const setSelectCard = getByTestId(TEST_ID.SELECT_CARD);
    fireEvent.click(setSelectCard);
    expect(setSelectCard).toBeInTheDocument();
  });

  test('Default credit card | it should be show icon', () => {
    const props = {
      ...props,
      isCanBeSelect: true,
      isCardSelected: true,
    };
    const { getByTestId } = render(<SavedCreditCardListItem {...props} />);
    const setDefaultCard = getByTestId(TEST_ID.DEFAULT_CARD);
    expect(setDefaultCard).toBeInTheDocument();
  });
});
