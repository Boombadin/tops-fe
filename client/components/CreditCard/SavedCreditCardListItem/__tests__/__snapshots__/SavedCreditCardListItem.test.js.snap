// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CreditCard/SavedCreditCardListItem it should match with snapshot 1`] = `
<DocumentFragment>
  <div
    class="SavedCreditCardListItem__CardBox-uqaej7-0 hxMYXb"
    data-testid="inf-viewSavedCreditCardListItem-Container"
  >
    <div
      class="SavedCreditCardListItem__CardBoxInside-uqaej7-1 dfbfSl"
      data-testid="btn-editSavedCreditCardListItem-OnSetSelectCard"
    >
      <div
        class="SavedCreditCardListItem__CardInfo-uqaej7-2 iGSYTl"
      >
        <div
          class="SavedCreditCardListItem__CardInfoBank-uqaej7-3 jLUkqG"
        >
          <div
            class="SavedCreditCardListItem__CardDefault-uqaej7-4 cZSUIs"
          />
          <div
            class="SavedCreditCardListItem__CardTypeIcon-uqaej7-6 bybrOI"
          >
            <img
              src="/icons/types-credit-card/master-no-border.svg"
            />
          </div>
          <p
            class="SavedCreditCardListItem__BankName-uqaej7-7 eTIArD Text-sc-9p67zt-0 gbahPG"
            color="default"
            data-testid="inf-viewSavedCreditCardListItem-BankName"
            display=""
            size="13"
            spacing="normal"
          />
          <p
            class="SavedCreditCardListItem__DefaultText-uqaej7-8 hxxdLp Text-sc-9p67zt-0 RCJvp"
            color="#b4b8c1"
            display=""
            size="13"
            spacing="normal"
          >
            payment.credit_card.card_list_default
          </p>
        </div>
        <div
          class="SavedCreditCardListItem__CardNumber-uqaej7-9 UHWxw"
        >
          * 4444
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
